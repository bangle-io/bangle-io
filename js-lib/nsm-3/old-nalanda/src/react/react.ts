import { useDebugValue, useRef, useState } from 'react';
import useSyncExternalStoreExports from 'use-sync-external-store/shim';

import type { Slice } from '../vanilla/slice';
import type { Store } from '../vanilla/store';
import type { AnySlice } from '../vanilla/types';

const { useSyncExternalStore } = useSyncExternalStoreExports;

type InferData<T> = T extends Slice<any, infer TData, any> ? TData : never;

interface ReactAdapter<TSnapshot> {
  subscribe: (onStoreChange: () => void) => () => void;
  getSnapshot: () => TSnapshot;
}

export function createUseTrackSliceHook<TAllSliceName extends string = any>(
  store: Store<TAllSliceName>,
) {
  function useSlice<TSlice extends AnySlice>(sl: TSlice): InferData<TSlice> {
    const ref = useRef<InferData<TSlice>>();

    const [adapter] = useState(() => {
      const adapter: ReactAdapter<InferData<TSlice>> = {
        subscribe: (onStoreChange) => {
          const sliceEffect = store.effect((effectStore) => {
            const selectedData = sl.track(effectStore);
            ref.current = selectedData;

            onStoreChange();

            queueMicrotask(() => {
              if (effectStore._runInstance?._addTrackedCount === 0) {
                console.warn(
                  `You forgot to destructure/access the tracked field from ${sl.name}. This will not track any changes!.`,
                );
              }
            });
          });

          return () => {
            store.unregisterEffect(sliceEffect);
          };
        },
        getSnapshot: () => {
          return ref.current ?? sl.get(store.state);
        },
      };

      return adapter;
    });

    const data = useSyncExternalStore(adapter.subscribe, adapter.getSnapshot);

    useDebugValue(data);

    return data;
  }

  return useSlice;
}
