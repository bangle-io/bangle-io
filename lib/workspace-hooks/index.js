export * from './WorkspaceHooksContext';
export * from './workspace-hooks';
