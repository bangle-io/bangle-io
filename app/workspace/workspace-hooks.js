import { useHistory, useLocation, useParams } from 'react-router-dom';
import { useEffect, useState, useCallback } from 'react';
import { uuid } from '@bangle.dev/core/utils/js-utils';
import { Node } from '@bangle.dev/core/prosemirror/model';
import { resolvePath } from 'bangle-io/app/workspace/path-helpers';
import {
  createWorkspace,
  deleteWorkspace,
  listWorkspaces,
} from './workspace-helpers';
import {
  createFile,
  deleteFile,
  listAllFiles,
  renameFile,
} from './file-helpers';
import { specRegistry } from '../editor/spec-sheet';

export function useGetWorkspaceFiles() {
  const { wsName } = useWorkspacePath();

  const [files, setFiles] = useState([]);

  const refreshFiles = useCallback(() => {
    if (wsName) {
      listAllFiles(wsName).then((items) => {
        setFiles(items);
      });
    }
  }, [wsName]);

  useEffect(() => {
    refreshFiles();
  }, [refreshFiles]);
  return [files, refreshFiles];
}

export function useCreateMdFile() {
  const { pushWsPath } = useWorkspacePath();

  const createNewMdFile = useCallback(
    async (
      wsPath,
      doc = Node.fromJSON(specRegistry.schema, {
        type: 'doc',
        content: [
          {
            type: 'heading',
            attrs: {
              level: 1,
            },
            content: [
              {
                type: 'text',
                text: resolvePath(wsPath).fileName,
              },
            ],
          },
          {
            type: 'paragraph',
            content: [
              {
                type: 'text',
                text: 'Hello world!',
              },
            ],
          },
        ],
      }),
    ) => {
      await createFile(wsPath, doc);
      pushWsPath(wsPath);
    },
    [pushWsPath],
  );

  return createNewMdFile;
}

export function useRenameActiveFile() {
  const { wsName, wsPath, replaceWsPath } = useWorkspacePath();

  const renameFileCb = useCallback(
    async (newFilePath) => {
      if (!newFilePath.endsWith('.md')) {
        newFilePath += '.md';
      }

      const newWsPath = wsName + ':' + newFilePath;
      await renameFile(wsPath, newWsPath);
      replaceWsPath(newWsPath);
    },
    [wsName, wsPath, replaceWsPath],
  );

  return renameFileCb;
}

export function useDeleteFile() {
  const { wsName, wsPath } = useWorkspacePath();
  const history = useHistory();

  const deleteByDocName = useCallback(
    async (wsPathToDelete) => {
      await deleteFile(wsPathToDelete);
      if (wsPathToDelete === wsPath) {
        history.replace('/ws/' + wsName);
      }
    },
    [wsName, wsPath, history],
  );

  return deleteByDocName;
}

export function useWorkspaces() {
  const [workspaces, updateWorkspaces] = useState([]);
  const history = useHistory();

  useEffect(() => {
    let destroyed = false;
    listWorkspaces().then((workspaces) => {
      if (!destroyed) {
        updateWorkspaces(workspaces);
      }
    });
    return () => {
      destroyed = true;
    };
  }, []);

  const createWorkspaceCb = useCallback(
    async (wsName, type, opts) => {
      await createWorkspace(wsName, type, opts);
      history.push(`/ws/${wsName}`);
    },
    [history],
  );

  const deleteWorkspaceCb = useCallback(
    async (wsName) => {
      await deleteWorkspace(wsName);
      history.push(`/ws/`);
    },
    [history],
  );

  return {
    workspaces,
    createWorkspace: createWorkspaceCb,
    deleteWorkspace: deleteWorkspaceCb,
  };
}

export function useWorkspacePath() {
  const { wsName } = useParams();
  const { pathname } = useLocation();
  const history = useHistory();

  const pushWsPath = useCallback(
    (wsPath) => {
      const { wsName, filePath } = resolvePath(wsPath);
      history.push(`/ws/${wsName}/${filePath}`);
    },
    [history],
  );

  const replaceWsPath = useCallback(
    (wsPath) => {
      const { wsName, filePath } = resolvePath(wsPath);
      history.replace(`/ws/${wsName}/${filePath}`);
    },
    [history],
  );

  if (!wsName) {
    return {};
  }

  const filePath = pathname.split('/').slice(3).join('/');
  let wsPath;
  let docName;

  if (filePath) {
    wsPath = wsName + ':' + filePath;
    ({ docName } = resolvePath(wsPath));
  }

  return {
    wsName,
    wsPath,
    docName,
    filePath,
    pushWsPath,
    replaceWsPath,
  };
}
