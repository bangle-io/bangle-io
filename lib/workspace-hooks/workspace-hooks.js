import { useState, useCallback, useEffect } from 'react';
import {
  useWorkspacePath,
  listAllFiles,
  isValidNoteWsPath,
} from 'workspace/index';
import { useHistory, matchPath, useLocation } from 'react-router-dom';
