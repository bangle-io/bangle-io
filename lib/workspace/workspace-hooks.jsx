import { useHistory, matchPath, useLocation } from 'react-router-dom';
import { useEffect, useState, useCallback } from 'react';
import { Node } from '@bangle.dev/core/prosemirror/model';
import {
  isValidNoteWsPath,
  locationToFilePath,
  resolvePath,
  toFSPath,
} from './path-helpers';
import {
  createWorkspace,
  deleteWorkspace,
  getWorkspaceInfo,
  listWorkspaces,
} from './workspace-helpers';
import { checkWidescreen, useDestroyRef } from 'utils/index';
import { importGithubWorkspace } from './github-helpers';
import {
  HELP_FS_INDEX_FILE_NAME,
  HELP_FS_WORKSPACE_NAME,
  HELP_FS_WORKSPACE_TYPE,
} from 'config/help-fs';
import { getFileSystemFromWsInfo } from './get-fs';
const LOG = false;
let log = LOG ? console.log.bind(console, 'workspace/index') : () => {};

export function useWorkspaces() {
  const [workspaces, updateWorkspaces] = useState([]);
  const { wsName } = useWorkspacePath();
  const history = useHistory();
  const destroyedRef = useDestroyRef();

  const refreshWorkspaces = useCallback(() => {
    listWorkspaces().then((workspaces) => {
      if (!destroyedRef.current) {
        updateWorkspaces(workspaces);
      }
    });
  }, [destroyedRef]);

  const createWorkspaceCb = useCallback(
    async (wsName, type, opts) => {
      await createWorkspace(wsName, type, opts);
      history.push(`/ws/${wsName}`);
    },
    [history],
  );

  const importWorkspaceFromGithubCb = useCallback(
    // can pass alternate wsName in the options
    async (bangleIOContext, url, wsType, opts = {}) => {
      const wsName = await importGithubWorkspace(
        bangleIOContext,
        url,
        wsType,
        opts.wsName,
        opts.token,
      );

      await refreshWorkspaces();
      history.push(`/ws/${wsName}`);
    },
    [history, refreshWorkspaces],
  );

  const deleteWorkspaceCb = useCallback(
    async (targetWsName) => {
      await deleteWorkspace(targetWsName);
      if (targetWsName === wsName) {
        history.push(`/ws/`);
      } else {
        refreshWorkspaces();
      }
    },
    [history, wsName, refreshWorkspaces],
  );

  useEffect(() => {
    refreshWorkspaces();
  }, [refreshWorkspaces]);

  const switchWorkspaceCb = useCallback(
    async (wsName, newTab) => {
      const newPath = '/ws/' + wsName;
      if (newTab) {
        window.open(newPath);
        return;
      }
      history.push(newPath);
    },
    [history],
  );

  return {
    workspaces,
    createWorkspace: createWorkspaceCb,
    deleteWorkspace: deleteWorkspaceCb,
    switchWorkspace: switchWorkspaceCb,
    importWorkspaceFromGithub: importWorkspaceFromGithubCb,
  };
}

/**
 * WARNING check if wsName, wsPath exists before working on it
 */
export function useWorkspacePath() {
  // I think history doesn't change when location changes
  // so it is a good idea to use useLocation instead of location
  const location = useLocation();
  const history = useHistory();
  const match = matchPath(location.pathname, {
    path: '/ws/:wsName',
    exact: false,
    strict: false,
  });

  const { wsName = HELP_FS_WORKSPACE_NAME } = match?.params ?? {};

  let wsPath;
  let filePath = locationToFilePath(location);
  if (filePath) {
    wsPath = wsName + ':' + filePath;
  }
  // show the landing page if no file path i.e. at root '/'
  else if (wsName === HELP_FS_WORKSPACE_NAME && !filePath) {
    wsPath = wsName + ':' + HELP_FS_INDEX_FILE_NAME;
    filePath = resolvePath(wsPath).filePath;
  }

  const search = new URLSearchParams(location?.search);
  const secondaryWsPath = search.get('secondary') ?? undefined;

  const replaceSecondaryWsPath = useCallback(
    (wsPath) => {
      const isWidescreen = checkWidescreen();
      const newSearch = new URLSearchParams(location?.search);
      if (isWidescreen) {
        // replace is intentional as native history pop
        // for some reason isnt remembering the state.
        if (isWidescreen) {
          newSearch.set('secondary', wsPath);
        } else {
          newSearch.delete('secondary');
        }
        history.replace({
          ...location,
          search: newSearch.toString(),
        });
        return;
      }
    },
    [history, location],
  );

  const pushWsPath = useCallback(
    (wsPath, newTab = false, secondary = false) => {
      const { wsName, filePath } = resolvePath(wsPath);
      const newPath = encodeURI(`/ws/${wsName}/${filePath}`);
      if (newTab) {
        window.open(newPath);
        return;
      }

      const isWidescreen = checkWidescreen();
      const newSearch = new URLSearchParams(location?.search);

      if (isWidescreen && secondary) {
        replaceSecondaryWsPath(wsPath);
        return;
      }

      if (newPath === location.pathname && !secondary) {
        return;
      }
      history.push({
        ...location,
        pathname: newPath,
        search: newSearch.toString(),
      });
    },
    [history, location, replaceSecondaryWsPath],
  );

  const replaceWsPath = useCallback(
    (wsPath) => {
      const { wsName, filePath } = resolvePath(wsPath);
      log('replaceWsPath', wsPath);

      history.replace({
        ...location,
        pathname: encodeURI(`/ws/${wsName}/${filePath}`),
      });
    },
    [history, location],
  );

  // removes the currently active wsPath
  const removeWsPath = useCallback(() => {
    if (!wsPath) {
      return;
    }

    let newPath = null;

    // transition any secondary to main
    if (secondaryWsPath) {
      const { wsName, filePath } = resolvePath(secondaryWsPath);
      newPath = encodeURI(`/ws/${wsName}/${filePath}`);
    } else {
      newPath = encodeURI(`/ws/${wsName}`);
    }

    history.push(newPath);
  }, [history, wsPath, wsName, secondaryWsPath]);

  // the editor on side
  const removeSecondaryWsPath = useCallback(() => {
    log('removeSecondaryWsPath');
    const newSearch = new URLSearchParams(location?.search);
    newSearch.delete('secondary');
    history.replace({
      ...location,
      search: newSearch.toString(),
    });
  }, [history, location]);

  const removePrimaryAndSecondaryWsPath = useCallback(() => {
    log('removePrimaryAndSecondaryWsPath');
    const newSearch = new URLSearchParams(location?.search);
    newSearch.delete('secondary');
    history.replace({
      ...location,
      pathname: encodeURI(`/ws/${wsName}`),
      search: newSearch.toString(),
    });
  }, [wsName, history, location]);

  const replacePrimaryAndSecondaryWsPath = useCallback(
    (primaryWsPath, secondaryWsPath) => {
      const { wsName, filePath } = resolvePath(primaryWsPath);
      const newPath = encodeURI(`/ws/${wsName}/${filePath}`);
      const isWidescreen = checkWidescreen();

      const newSearch = new URLSearchParams(location?.search);
      if (secondaryWsPath) {
        if (isWidescreen) {
          newSearch.set('secondary', secondaryWsPath);
        } else {
          newSearch.delete('secondary');
        }
      }

      history.push({
        ...location,
        pathname: newPath,
        search: newSearch.toString(),
      });
    },
    [history, location],
  );

  return {
    wsName,
    wsPath,
    secondaryWsPath,
    filePath,
    pushWsPath,
    replaceWsPath,
    replaceSecondaryWsPath,
    removeWsPath,
    removeSecondaryWsPath,
    removePrimaryAndSecondaryWsPath,
    replacePrimaryAndSecondaryWsPath,
  };
}

const isHelpFileModified = async (wsPath) => {
  if (!wsPath) {
    return false;
  }
  const { wsName } = resolvePath(wsPath);

  if (wsName !== HELP_FS_WORKSPACE_NAME) {
    return false;
  }

  const wsInfo = await getWorkspaceInfo(wsName);

  if (wsInfo.type !== HELP_FS_WORKSPACE_TYPE) {
    return false;
  }

  const fs = getFileSystemFromWsInfo(wsInfo);

  const isModified = await fs.isFileModified(toFSPath(wsPath));

  return isModified;
};

/**
 * A hook that checks if the user has made any modifications
 * to the help workspace. Will return isModified false is not currently
 * in help workspace.
 * @param {*} checkInterval time to check
 * @returns
 */
export function useIsHelpWorkspaceModified(wsPath, checkInterval = 6000) {
  const [isModified, updateModified] = useState(false);

  const wsName = wsPath && resolvePath(wsPath).wsName;

  const isHelpWorkspace = wsName === HELP_FS_WORKSPACE_NAME;

  useEffect(() => {
    let id;
    let effectDestroyed = false;
    if (isHelpWorkspace && wsPath && isValidNoteWsPath(wsPath)) {
      id = setInterval(() => {
        isHelpFileModified(wsPath).then((result) => {
          if (effectDestroyed) {
            return;
          }
          if (result) {
            updateModified(result);
          }
        });
      }, checkInterval);
    }
    return () => {
      effectDestroyed = true;
      if (id != null) {
        clearInterval(id);
      }
    };
  }, [wsPath, wsName, isModified, checkInterval, isHelpWorkspace]);

  useEffect(() => {
    return () => {
      updateModified(false);
    };
  }, [wsName]);

  return { isModified, isHelpWorkspace };
}
