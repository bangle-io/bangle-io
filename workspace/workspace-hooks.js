import { useHistory, matchPath, useLocation } from 'react-router-dom';
import { useEffect, useState, useCallback, useRef } from 'react';
import { Node } from '@bangle.dev/core/prosemirror/model';
import { specRegistry } from 'editor/index';

import { locationToFilePath, resolvePath } from './path-helpers';
import {
  createWorkspace,
  deleteWorkspace,
  listWorkspaces,
} from './workspace-helpers';
import {
  createFile,
  deleteFile,
  listAllFiles,
  renameFile,
} from './file-helpers';
import { checkWidescreen } from 'utils/index';
import { importGithubWorkspace } from './github-helpers';
import { FILE_NOT_FOUND_ERROR } from 'baby-fs';

const LOG = false;
let log = LOG ? console.log.bind(console, 'workspace/index') : () => {};

export function useGetWorkspaceFiles() {
  const { wsName, wsPermissionState } = useWorkspacePath();

  const [files, setFiles] = useState([]);
  const isDestroyed = useRef(false);

  const refreshFiles = useCallback(() => {
    if (
      wsName &&
      (wsPermissionState.type === 'ready' ||
        // TODO I am not happy with this error code check here
        // I feel this part of code shouldnt know so much about error and codes
        // Can we make the wsPermissionState accomodate the common 404 not found error.
        // allow listing of files if the current file is not found
        wsPermissionState.error?.code === FILE_NOT_FOUND_ERROR)
    ) {
      // TODO this is called like a million times
      // we need to fix this to only update based on known things
      // like renaming of file, delete etc.
      listAllFiles(wsName).then((items) => {
        if (!isDestroyed.current) {
          setFiles(items);
        }
      });
    }
    // TODO: look into this the one below is a quick fix as stale files are shown
    // when you switch a workspace but waiting on the permission page.
    if (wsName && wsPermissionState.type === 'permission') {
      if (!isDestroyed.current) {
        setFiles([]);
      }
    }
  }, [wsName, wsPermissionState]);

  useEffect(() => {
    refreshFiles();
  }, [refreshFiles]);

  useEffect(() => {
    return () => {
      isDestroyed.current = true;
    };
  }, []);

  return [files, refreshFiles];
}

export function useCreateMdFile() {
  const { pushWsPath } = useWorkspacePath();

  const createNewMdFile = useCallback(
    async (
      wsPath,
      doc = Node.fromJSON(specRegistry.schema, {
        type: 'doc',
        content: [
          {
            type: 'heading',
            attrs: {
              level: 1,
            },
            content: [
              {
                type: 'text',
                text: resolvePath(wsPath).fileName,
              },
            ],
          },
          {
            type: 'paragraph',
            content: [
              {
                type: 'text',
                text: 'Hello world!',
              },
            ],
          },
        ],
      }),
    ) => {
      await createFile(wsPath, doc);
      pushWsPath(wsPath);
    },
    [pushWsPath],
  );

  return createNewMdFile;
}

export function useRenameActiveFile() {
  const { wsName, wsPath, replaceWsPath } = useWorkspacePath();

  const renameFileCb = useCallback(
    async (newFilePath) => {
      if (!newFilePath.endsWith('.md')) {
        newFilePath += '.md';
      }

      const newWsPath = wsName + ':' + newFilePath;
      await renameFile(wsPath, newWsPath);
      replaceWsPath(newWsPath);
    },
    [wsName, wsPath, replaceWsPath],
  );

  return renameFileCb;
}

export function useDeleteFile() {
  const { wsName, wsPath } = useWorkspacePath();
  const history = useHistory();

  const deleteByWsPath = useCallback(
    async (wsPathToDelete) => {
      await deleteFile(wsPathToDelete);
      if (wsPathToDelete === wsPath) {
        history.replace('/ws/' + wsName);
      }
    },
    [wsName, wsPath, history],
  );

  return deleteByWsPath;
}

export function useWorkspaces() {
  const [workspaces, updateWorkspaces] = useState([]);
  const { wsName } = useWorkspacePath();
  const history = useHistory();
  const isDestroyed = useRef(false);

  const refreshWorkspaces = useCallback(() => {
    listWorkspaces().then((workspaces) => {
      if (!isDestroyed.current) {
        updateWorkspaces(workspaces);
      }
    });
  }, []);

  useEffect(() => {
    refreshWorkspaces();
    return () => {
      isDestroyed.current = true;
    };
  }, [refreshWorkspaces]);

  const createWorkspaceCb = useCallback(
    async (wsName, type, opts) => {
      await createWorkspace(wsName, type, opts);
      history.push(`/ws/${wsName}`);
    },
    [history],
  );

  const importWorkspaceFromGithubCb = useCallback(
    // can pass alternat wsName in the options
    async (url, wsType, opts = {}) => {
      const wsName = await importGithubWorkspace(
        url,
        wsType,
        opts.wsName,
        opts.token,
      );

      await refreshWorkspaces();
      history.push(`/ws/${wsName}`);
    },
    [history, refreshWorkspaces],
  );

  const deleteWorkspaceCb = useCallback(
    async (targetWsName) => {
      await deleteWorkspace(targetWsName);
      if (targetWsName === wsName) {
        history.push(`/ws/`);
      } else {
        refreshWorkspaces();
      }
    },
    [history, wsName, refreshWorkspaces],
  );

  const switchWorkspaceCb = useCallback(
    async (wsName, newTab) => {
      const newPath = '/ws/' + wsName;
      if (newTab) {
        window.open(newPath);
        return;
      }
      history.push(newPath);
    },
    [history],
  );

  return {
    workspaces,
    createWorkspace: createWorkspaceCb,
    deleteWorkspace: deleteWorkspaceCb,
    switchWorkspace: switchWorkspaceCb,
    importWorkspaceFromGithub: importWorkspaceFromGithubCb,
  };
}

export function useWorkspacePath() {
  const location = useLocation();
  const match = matchPath(location.pathname, {
    path: '/ws/:wsName',
    exact: false,
    strict: false,
  });

  const { wsName } = match?.params ?? {};
  const { secondaryWsPath } = location?.state ?? {};
  let wsPath;

  const filePath = locationToFilePath(location);
  if (filePath) {
    wsPath = wsName + ':' + filePath;
  }

  const history = useHistory();

  const pushWsPath = useCallback(
    (wsPath, newTab = false, secondary = false) => {
      const { wsName, filePath } = resolvePath(wsPath);
      const newPath = `/ws/${wsName}/${filePath}`;

      if (newTab) {
        window.open(newPath);
        return;
      }

      const isWidescreen = checkWidescreen();

      if (isWidescreen && secondary) {
        // replace is intentional as native history pop
        // for some reason isnt remembering the state.
        history.replace(history.location.pathname, {
          ...history.location.state,
          secondaryWsPath: wsPath,
        });
        return;
      }

      if (newPath === history.location.pathname) {
        return;
      }

      history.push(newPath, {
        ...history.location.state,
        secondaryWsPath: isWidescreen ? secondaryWsPath : null,
      });
    },
    [history, secondaryWsPath],
  );

  const replaceWsPath = useCallback(
    (wsPath) => {
      const { wsName, filePath } = resolvePath(wsPath);
      log('replaceWsPath', wsPath);

      history.replace({
        ...history.location,
        pathname: `/ws/${wsName}/${filePath}`,
      });
    },
    [history],
  );

  const setWsPermissionState = useCallback(
    (payload) => {
      log(
        'setWsPermissionState',
        payload,
        'existing state',
        history.location.state,
      );
      history.replace({
        ...history.location,
        state: { ...history.location.state, wsPermissionState: payload },
      });
    },
    [history],
  );

  // removes the currently active wsPath
  const removeWsPath = useCallback(() => {
    if (!wsPath) {
      return;
    }

    let newPath = null;

    // transition any secondary to main
    if (secondaryWsPath) {
      const { wsName, filePath } = resolvePath(secondaryWsPath);
      newPath = `/ws/${wsName}/${filePath}`;
    }

    history.push(newPath, {
      ...history.location.state,
      secondaryWsPath: null,
    });
  }, [history, wsPath, secondaryWsPath]);

  // the editor on side
  const removeSecondaryWsPath = useCallback(() => {
    log('removeSecondaryWsPath');
    history.replace(history.location.pathname, {
      ...history.location.state,
      secondaryWsPath: null,
    });
  }, [history]);

  // TODO should I add more safeguard for
  // workspacePerm.type == ready?
  if (!wsName) {
    return {
      wsName,
      wsPath: null,
      secondaryWsPath: null,
      filePath: null,
      pushWsPath,
      replaceWsPath,
      setWsPermissionState,
      removeSecondaryWsPath,
      removeWsPath,
      wsPermissionState: {},
    };
  }

  return {
    wsName,
    wsPath,
    secondaryWsPath,
    filePath,
    pushWsPath,
    replaceWsPath,
    setWsPermissionState,
    removeWsPath,
    removeSecondaryWsPath,
    wsPermissionState: location?.state?.wsPermissionState ?? {},
  };
}
