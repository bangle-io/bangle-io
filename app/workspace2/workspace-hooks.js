import { useHistory, useLocation, useParams } from 'react-router-dom';
import { useContext, useEffect, useState, useCallback } from 'react';
import { uuid } from '@bangle.dev/core/utils/js-utils';

import { EditorManagerContext } from './EditorManager';
import { requestPermission as requestFilePermission } from '../workspace/native-fs-driver';
import {
  createFile,
  createWorkspace,
  deleteFile,
  getFiles,
  getWorkspaceInfo,
  getWorkspaces,
  resolvePath,
  wsQueryPermission,
} from './workspace-helpers';

export function useGetWorkspaceFiles() {
  const { wsName } = useWorkspaceDetails();

  const [files, setFiles] = useState([]);

  const refreshFiles = useCallback(() => {
    if (wsName) {
      getFiles(wsName).then((items) => {
        setFiles(items);
      });
    }
  }, [wsName]);

  useEffect(() => {
    refreshFiles();
  }, [refreshFiles]);
  return [files, refreshFiles];
}

// TODO does it really need to be a hook
export function useCreateNewFile() {
  const { wsName, pushWsPath } = useWorkspaceDetails();

  const createNewFile = useCallback(
    async (fileName = uuid(6)) => {
      const wsPath = wsName + ':' + fileName;
      await createFile(wsPath);
      // const workspace = await getWorkspace(wsName);
      // const newFile = await workspace.createFile(wsPath, null);
      // workspace.linkFile(newFile);
      pushWsPath(wsPath);
    },
    [wsName, pushWsPath],
  );

  return createNewFile;
}

export function useDeleteByWsPath() {
  const { wsName } = useWorkspaceDetails();
  const history = useHistory();

  const deleteByDocName = useCallback(
    async (wsPath) => {
      await deleteFile(wsPath);
      history.push('/ws/' + wsName);
    },
    [wsName, history],
  );

  return deleteByDocName;
}

export function useWorkspaces() {
  const [workspaces, updateWorkspaces] = useState([]);
  const history = useHistory();

  useEffect(() => {
    let destroyed = false;
    getWorkspaces().then((workspaces) => {
      if (!destroyed) {
        updateWorkspaces(workspaces.map((w) => w.name));
      }
    });
    return () => {
      destroyed = true;
    };
  }, []);

  const createWorkspaceCb = useCallback(
    async (wsName, type) => {
      await createWorkspace(wsName, type);
      history.push(`/ws/${wsName}`);
    },
    [history],
  );

  return {
    workspaces,
    createWorkspace: createWorkspaceCb,
  };
}

export function useWorkspacePermission() {
  const {
    editorManagerState: { wsPermission: permission },
    dispatch,
  } = useContext(EditorManagerContext);

  const { wsName } = useWorkspaceDetails();

  const setPermission = useCallback(
    (value) => {
      dispatch({
        type: 'WORKSPACE/PERMISSION',
        value,
      });
    },
    [dispatch],
  );

  useEffect(() => {
    setPermission(undefined);
    wsQueryPermission(wsName).then((result) => {
      setPermission(result ? 'granted' : 'rejected');
    });
  }, [setPermission, wsName]);

  const requestPermission = useCallback(async () => {
    if (permission === 'granted') {
      return true;
    }
    const workspaceInfo = await getWorkspaceInfo(wsName);
    if (!workspaceInfo.metadata.dirHandle) {
      setPermission('granted');
      return true;
    }
    const isGranted = await requestFilePermission(
      workspaceInfo.metadata.dirHandle,
    );
    if (isGranted) {
      setPermission('granted');
      return true;
    } else {
      setPermission('rejected');
      return false;
    }
  }, [wsName, permission, setPermission]);

  return [permission, requestPermission];
}

export function useWorkspaceDetails() {
  const { wsName } = useParams();
  const { pathname } = useLocation();
  const history = useHistory();
  const pushWsPath = useCallback(
    (wsPath) => {
      const { wsName, filePath } = resolvePath(wsPath);
      history.push(`/ws/${wsName}/${filePath}`);
    },
    [history],
  );

  if (!wsName) {
    return {};
  }

  const filePath = pathname.split('/').slice(3).join('/');
  let wsPath;
  let docName;

  if (filePath) {
    wsPath = wsName + ':' + filePath;
    ({ docName } = resolvePath(wsPath));
  }

  return {
    wsName,
    wsPath,
    docName,
    filePath,
    pushWsPath,
  };
}
