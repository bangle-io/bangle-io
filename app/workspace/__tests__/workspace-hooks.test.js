import React from 'react';
import { MemoryRouter as Router, Switch, Route } from 'react-router-dom';

import { useCreateMdFile } from '../workspace-hooks';
import { render, act } from '@testing-library/react';
import * as idb from 'idb-keyval';

jest.mock('idb-keyval', () => {
  return {
    get: jest.fn(),
    del: jest.fn(),
    set: jest.fn(),
    keys: jest.fn(),
  };
});

describe('useCreateMdFile', () => {
  test('browser create file', async () => {
    let callback;

    idb.get.mockImplementation((key) => {
      if (key === 'workspaces/2') {
        return [
          {
            name: 'kujo',
            type: 'browser',
            metadata: {},
          },
        ];
      }
    });

    function Comp() {
      const createMdFile = useCreateMdFile();
      callback = createMdFile;
      return <div>Hello</div>;
    }

    await render(
      <Router initialEntries={['/ws/kujo']}>
        <Switch>
          <Route path={['/ws/:wsName']}>
            <Comp />
          </Route>
        </Switch>
      </Router>,
    );

    expect(callback).not.toBeUndefined();

    const result = callback('kujo:one.md');
    expect(result).toBeInstanceOf(Promise);

    await result;

    expect(idb.set).toBeCalledTimes(1);
    expect(idb.set).toBeCalledWith('kujo:one.md', {
      doc: {
        content: [
          {
            attrs: { level: 1 },
            content: [{ text: 'one.md', type: 'text' }],
            type: 'heading',
          },
          {
            content: [{ text: 'Hello world!', type: 'text' }],
            type: 'paragraph',
          },
        ],
        type: 'doc',
      },
      name: 'one.md',
    });
  });
});
