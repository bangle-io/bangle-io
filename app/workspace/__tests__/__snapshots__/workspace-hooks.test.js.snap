// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWorkspacePath refreshes after workspace permissions are updated 1`] = `
<div>
  <span>
    Encountered an Error
  </span>
  <br />
  <span>
    Error
    : 
    some magical error
  </span>
  <br />
  <span>
    Please try reloading the page or open a ticket if the issue persists
  </span>
</div>
`;
